packages/nitro-protocol/contracts/examples/HashLock.sol

```diff
@@ -0,0 +1,85 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.7.4;
+pragma experimental ABIEncoderV2;
+
+import '../interfaces/IForceMoveApp.sol';
+import '../Outcome.sol';
+
+/**
+ * @dev The HashLock contract complies with the ForceMoveApp interface and implements a hashlocked payment
+ */
+contract HashLock is IForceMoveApp {
+    struct AppData {
+        bytes32 h;
+        bytes preImage;
+    }
+
+    function validTransition(
+        VariablePart memory a,
+        VariablePart memory b,
+        uint48 turnNumB,
+        uint256
+    ) public override pure returns (bool) {
+        // is this the first and only swap?
+        require(turnNumB == 4, 'turnNumB != 4');
+
+        // Decode variables.
+        // Assumptions:
+        //  - single asset in this channel
+        //  - two parties in this channel
+        //  - not a "guarantee" channel (c.f. Nitro paper)
+        Outcome.AllocationItem[] memory allocationA = decode2PartyAllocation(a.outcome);
+        Outcome.AllocationItem[] memory allocationB = decode2PartyAllocation(b.outcome);
+        bytes memory preImage = abi.decode(b.appData, (AppData)).preImage;
+        bytes32 h = abi.decode(a.appData, (AppData)).h;
+
+        // is the preimage correct?
+        require(keccak256(preImage) == h, 'Incorrect preimage');
+
+        // slots for each participant unchanged
+        require(
+            allocationA[0].destination == allocationB[0].destination &&
+                allocationA[1].destination == allocationB[1].destination,
+            'destinations may not change'
+        );
+
+        // was the payment made?
+        require(
+            allocationA[0].amount == allocationB[1].amount &&
+                allocationA[1].amount == allocationB[0].amount,
+            'amounts must be permuted'
```

Comment:
This key property should probably be reflected in the name of the contract: e.g. `HashLockedAllocationSwap`.

Reply:
2260c14